src/relentless/simulate/hoomd.py

```diff
@@ -28,45 +45,102 @@
 
-    This is an abstract base class that needs to have its :meth:`initialize`
-    method implemented.
+    def _call_v3(self, sim):
+        raise NotImplementedError(
+            f"{self.__class__.__name__} not implemented in HOOMD 3"
+        )
 
-    """
+    def _call_v2(self, sim):
+        raise NotImplementedError(
+            f"{self.__class__.__name__} not implemented in HOOMD 2"
+        )
 
-    def __call__(self, sim):
-        # initialize hoomd exec conf once
-        if hoomd.context.exec_conf is None:
-            hoomd.context.initialize("--notice-level=0")
-            hoomd.util.quiet_status()
 
+class AnalysisOperation(simulate.AnalysisOperation):
+    def pre_run(self, sim, sim_op):
+        if _version.major == 3:
+            self._pre_run_v3(sim, sim_op)
+        elif _version.major == 2:
+            self._pre_run_v2(sim, sim_op)
+        else:
+            raise NotImplementedError(f"HOOMD {_version} not supported")
+
+    def post_run(self, sim, sim_op):
+        if _version.major == 3:
+            self._post_run_v3(sim, sim_op)
+        elif _version.major == 2:
+            self._post_run_v2(sim, sim_op)
+        else:
+            raise NotImplementedError(f"HOOMD {_version} not supported")
+
+    def _pre_run_v3(self, sim, sim_op):
+        raise NotImplementedError(
+            f"{self.__class__.__name__} not implemented in HOOMD 3"
+        )
+
+    def _pre_run_v2(self, sim, sim_op):
+        raise NotImplementedError(
+            f"{self.__class__.__name__} not implemented in HOOMD 2"
+        )
+
+    def _post_run_v3(self, sim, sim_op):
+        raise NotImplementedError(
+            f"{self.__class__.__name__} not implemented in HOOMD 3"
+        )
+
+    def _post_run_v2(self, sim, sim_op):
+        raise NotImplementedError(
+            f"{self.__class__.__name__} not implemented in HOOMD 2"
+        )
+
+
+# initializers
+class InitializationOperation(SimulationOperation, simulate.InitializationOperation):
+    def _call_v3(self, sim):
+        self._initialize_v3(sim)
+        sim.dimension = sim["engine"]["_hoomd"].state.box.dimensions
+        sim.types = sim["engine"]["_hoomd"].state.particle_types
+
+        # parse masses by type
+        snap = sim["engine"]["_hoomd"].state.get_snapshot()
+        sim.masses = self._get_masses_from_snapshot(sim, snap)
+
+        # create the potentials, defer attaching until later
+        neighbor_list = hoomd.md.nlist.Tree(buffer=sim.potentials.pair.neighbor_buffer)
+        pair_potential = hoomd.md.pair.Table(nlist=neighbor_list)
+        for i, j in sim.pairs:
+            r = sim.potentials.pair.x
+            u = sim.potentials.pair.energy((i, j))
+            f = sim.potentials.pair.force((i, j))
+            if numpy.any(numpy.isinf(u)) or numpy.any(numpy.isinf(f)):
+                raise ValueError("Pair potential/force is infinite at evaluated r")
+            pair_potential.params[(i, j)] = dict(r_min=r[0], U=u, F=f)
```

Comment:
For HOOMD 3 this needs to be: 
```
pair_potential.params[(i, j)] = dict(r_min=r[0], U=u[:-1], F=f[:-1])
pair_potential.r_cut[(i, j)] = r[-1]
```
We need to check that this is most consistent with the way HOOMD 2 works.

Reply:
This should be fixed. If everything seems to be working with the change applied, can you approve the PR? We'll merge and deal with the LAMMPS table issue, and trimming the `r_cut` down in a separate PR.